File: rugs/starwars.js
var starwars = new DuderRug("Star Wars", "Get information about Star Wars characters and locations.");

starwars.addCommand("starwars", function(cmd) {
    if (cmd.args.length < 3) {
        cmd.replyToAuthor("Usage: `starwars <character|location> <name>`");
        return;
    }

    var type = cmd.args[1].toLowerCase();
    var query = cmd.args.slice(2).join(" ");
    var url = "https://starwars-databank-server.vercel.app/api/v1/";

    Duder.startTyping(cmd.channelID);

    if (type === "character") {
        url += "characters/search/" + encodeURIComponent(query);
    } else if (type === "location") {
        url += "locations/search/" + encodeURIComponent(query);
    } else {
        cmd.replyToAuthor("Invalid search type. Use `character` or `location`.");
        return;
    }

    var content = HTTP.get(10, url, {});
    if (content === false) {
        cmd.replyToAuthor("Something went wrong while searching.");
        return;
    }

    // Check if the response is HTML by looking at the first character
    if (content.trim().substring(0, 1) === "<") {
        cmd.replyToAuthor("The Star Wars API returned an unexpected error. Please try again later.");
        this.wprint("API returned HTML instead of JSON from URL: " + url);
        return;
    }

    var json;
    try {
        json = JSON.parse(content);
    } catch (e) {
        cmd.replyToAuthor("There was an error parsing the data from the Star Wars API.");
        this.wprint("Failed to parse JSON: " + e.message);
        return;
    }

    if (type === "character") {
        if (json === undefined || !json.characters || json.characters.length === 0) {
            cmd.replyToAuthor("Could not find a character with that name.");
            return;
        }

        var character = json.characters[0];

        var embed = new EmbedMessage();
        embed.setTitle(character.name);
        embed.setDescription(character.description);
        embed.setThumbnail(character.image);
        embed.addField("Homeworld", character.homeworld || "Unknown");
        embed.addField("Species", character.species || "Unknown");
        embed.addField("Affiliations", character.affiliations ? character.affiliations.join(", ") : "None");
        embed.addField("Masters", character.masters ? character.masters.join(", ") : "None");
        embed.addField("Apprentices", character.apprentices ? character.apprentices.join(", ") : "None");

        cmd.replyToChannelEmbed(embed.compile());

    } else if (type === "location") {
        if (json === undefined || !json.locations || json.locations.length === 0) {
            cmd.replyToAuthor("Could not find a location with that name.");
            return;
        }

        var location = json.locations[0];

        var embed = new EmbedMessage();
        embed.setTitle(location.name);
        embed.setDescription(location.description);
        embed.setThumbnail(location.image);
        embed.addField("Climate", location.climate || "Unknown");
        embed.addField("Terrain", location.terrain || "Unknown");
        embed.addField("Notable Inhabitants", location.notable_inhabitants ? location.notable_inhabitants.join(", ") : "None");

        cmd.replyToChannelEmbed(embed.compile());
    }
});

File: rugs/starwars.go
package rugs

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/bwmarrin/discordgo"
)

func init() {
	Register(&StarWarsCommand{})
}

// StarWarsCommand implements the /starwars slash command
type StarWarsCommand struct{}

const starWarsAPIBase = "https://starwars-databank-server.vercel.app/api/v1"

func (c *StarWarsCommand) Name() string {
	return "starwars"
}

func (c *StarWarsCommand) Description() string {
	return "Search the Star Wars databank"
}

func (c *StarWarsCommand) Options() []*discordgo.ApplicationCommandOption {
	return []*discordgo.ApplicationCommandOption{
		{
			Type:        discordgo.ApplicationCommandOptionString,
			Name:        "type",
			Description: "What to search for",
			Required:    true,
			Choices: []*discordgo.ApplicationCommandOptionChoice{
				{Name: "Character", Value: "characters"},
				{Name: "Planet", Value: "locations"},
				{Name: "Vehicle/Starship", Value: "vehicles"},
				{Name: "Species", Value: "species"},
				{Name: "Droid", Value: "droids"},
				{Name: "Organization", Value: "organizations"},
			},
		},
		{
			Type:        discordgo.ApplicationCommandOptionString,
			Name:        "name",
			Description: "Name to search for",
			Required:    true,
		},
	}
}

func (c *StarWarsCommand) Execute(ctx CommandContext) error {
	searchType := ctx.GetString("type")
	name := ctx.GetString("name")

	ctx.DeferReply()

	// Fetch all items for the category (limit=1000 should cover everything)
	url := fmt.Sprintf("%s/%s?page=1&limit=1000", starWarsAPIBase, searchType)
	resp, err := ctx.HTTPGetString(10, url, nil)
	if err != nil {
		return ctx.FollowUp("Failed to access Star Wars databank.")
	}

	// Generic structure for all Databank items
	type DatabankItem struct {
		Name        string `json:"name"`
		Description string `json:"description"`
		Image       string `json:"image"`
	}

	var result struct {
		Data []DatabankItem `json:"data"`
	}

	if err := json.Unmarshal([]byte(resp), &result); err != nil {
		return ctx.FollowUp("Failed to parse databank response.")
	}

	// Fuzzy search
	var match *DatabankItem
	nameLower := strings.ToLower(name)

	// 1. Exact match
	for _, item := range result.Data {
		if strings.ToLower(item.Name) == nameLower {
			match = &item
			break
		}
	}

	// 2. Contains match (if no exact match)
	if match == nil {
		for _, item := range result.Data {
			if strings.Contains(strings.ToLower(item.Name), nameLower) {
				match = &item
				break
			}
		}
	}

	if match == nil {
		return ctx.FollowUp(fmt.Sprintf("No results found for '%s' in %s.", name, searchType))
	}

	embed := NewEmbed().
		SetTitle(match.Name).
		SetDescription(match.Description).
		SetImage(match.Image).
		SetColor(ColorGold).
		Build()

	return ctx.FollowUpEmbed(embed)
}

// Remove unused handlers
